pageserver/src/tenant/vectored_blob_io.rs

```diff
@@ -274,6 +275,7 @@ impl<'a> VectoredBlobReader<'a> {
         &self,
         read: &VectoredRead,
         buf: BytesMut,
+        new_storage_format: bool,
```

Comment:
I do not think that we should treate storage format as enum: it is just number.
In other Neon places where we need to maintain format version (i.e. metadata) it is just an integer.

I prefer to keep all logic with handling storage format version in image_layer.rs and pass bool to vectored_blob_io.
I agree that `new_storage_format` is bad name, I renamed it to `compressed_storage_format`.

Reply:
>In other Neon places where we need to maintain format version (i.e. metadata) it is just an integer.

I consider that a bug rather than a feature -- we should fix those as we evolve the code.  It's important that we don't risk bugs where someone passes some interchangable u32s in the wrong order: strong typing addresses this.

>I prefer to keep all logic with handling storage format version in image_layer.rs and pass bool to vectored_blob_io.

I think it makes sense to have a pub-crate enum of storage versions: then you don't need a new type for this function: you can just pass in the version and this function can do whatever the right thing is for the version in use.

Reply:
Still can not agree with you here.
Version (of storage format, metadata, product, ...) is integer or even string. In some cases (like in Postgres) it is splitted in major/minor parts. But I do not see any reasons to make it enum. There is IMHP completely no sense to try to assign some names to format version. It is just a number: 3.

Concerning strict type checking performed by compiler, it is also not the case: storage version is extracted and checked internally. It is not a part of high level API.

Reply:
We do not have 4 billion versions: that's  why a `u32` is not a good type for expressing the version.

Enums are valuable quality tools:
- they ensure that any place we're doing a `match` on the version, if we ever add a new version, we have to update that `match` or the code won't compile.
- they ensure that when decoding from persistent storage, if we see a version we don't understand, we won't try and do something with it.  This is important because of evolving versions (i.e. we shouldn't try and decode future-format data), but it is also important because a code bug like reading the wrong byte during decode can be caught if we only tolerate known versions.

Reply:
>We have just one version: current.

This thread is about the reader code.  We do not just have one version: we have to read whichever versions are present in on-disk data.

The starting point for this thread is the use of a `bool` to tell the function which version it's reading: the "before this PR" format or the "after this PR format".

What I'm saying is that rather than having a function prototype where the reader has to know about this format change and understand that the bool means "before PR #7091" or "after PR #7091", it would be much easier to understand and maintain.

To be clear, I'm not asking for some large structural change: you basically just:
- define an enum
- make the code that loads layer version throw an error if the version is unknown (we should never try to load data for a version we don't understand)
- pass that version around into reader objects rather than adding bool flags.

```
#[repr(u16)]
enum FormatVersion {
// Historical: versions prior to 3 date from 2021/2022, and no persistent pageserver data is still in use from that era.
V3 = 3, // Legacy format: uncompressed layers
V4 = 4 // Image layer compression is added in this version
}
```